File: dayz-community-hub-ui/src/lib/actions/dayzavr.ts
import { invoke, Channel } from '@tauri-apps/api/core';
import type { DayzavrServer } from '$lib/types';
import { app as s } from '$lib/state.svelte';
import * as m from '$lib/paraglide/messages.js';
import { saveDayzavrSettings } from './profile';

/** Fetch the DayZavr community server list (cached in state). */
export async function loadDayzavrServers(force = false) {
  if (s.dayzavrLoading) return;
  if (!force && s.dayzavrServers.length > 0) return;
  s.dayzavrLoading = true;
  try {
    s.dayzavrServers = await invoke<DayzavrServer[]>('fetch_dayzavr_servers');
  } catch (e) {
    s.setStatus(m.dayzavr_load_failed({ error: String(e) }), 'error');
  } finally {
    s.dayzavrLoading = false;
  }
}

/** Progress event streamed from the Rust torrent installer. */
interface DayzavrInstallProgress {
  status: string;
  downloaded_bytes: number;
  total_bytes: number;
  uploaded_bytes: number;
  download_mbps: number;
  upload_mbps: number;
  peers_live: number;
  peers_connecting: number;
  peers_seen: number;
  eta: string | null;
  done: boolean;
  error: string | null;
}

/** Install/update the mods a DayZavr server requires into the configured DayZ path. */
export async function installDayzavrMods(server: DayzavrServer) {
  const dayzPath = s.profile?.dayzavr_dayz_path;
  if (!dayzPath) {
    s.setStatus(m.dayzavr_set_path_first(), 'warning');
    return;
  }
  if (s.dayzavrInstall?.active) return;

  s.dayzavrInstall = {
    active: true,
    serverName: server.name,
    status: m.dayzavr_install_starting(),
    downloadedBytes: 0,
    totalBytes: 0,
    uploadedBytes: 0,
    downloadMbps: 0,
    uploadMbps: 0,
    peersLive: 0,
    peersConnecting: 0,
    peersSeen: 0,
    eta: null,
    done: false,
    error: null,
  };

  const onProgress = new Channel<DayzavrInstallProgress>();
  onProgress.onmessage = (p) => {
    if (!s.dayzavrInstall) return;
    s.dayzavrInstall = {
      ...s.dayzavrInstall,
      status: p.status,
      downloadedBytes: p.downloaded_bytes,
      totalBytes: p.total_bytes,
      uploadedBytes: p.uploaded_bytes,
      downloadMbps: p.download_mbps,
      uploadMbps: p.upload_mbps,
      peersLive: p.peers_live,
      peersConnecting: p.peers_connecting,
      peersSeen: p.peers_seen,
      eta: p.eta,
      done: p.done,
      error: p.error,
      active: !p.done && !p.error,
    };
    // Refresh installed set when a download finishes successfully.
    if (p.done && !p.error) loadDayzavrInstalledMods();
  };

  try {
    await invoke('install_dayzavr_mods', { mods: server.mods, dayzPath, onProgress });
  } catch (e) {
    if (s.dayzavrInstall) {
      s.dayzavrInstall = { ...s.dayzavrInstall, active: false, error: String(e) };
    }
  }
}

/** Refresh the set of fully-installed DayZavr mods (gates the Play button). */
export async function loadDayzavrInstalledMods() {
  const dayzPath = s.profile?.dayzavr_dayz_path;
  if (!dayzPath) {
    s.dayzavrInstalledMods = [];
    return;
  }
  try {
    s.dayzavrInstalledMods = await invoke<string[]>('dayzavr_installed_mods', { dayzPath });
  } catch {
    /* non-fatal (e.g. offline) */
  }
}

/** Whether every mod a server requires is installed (so it can be joined). */
export function serverModsInstalled(server: DayzavrServer): boolean {
  const set = new Set(s.dayzavrInstalledMods);
  return server.mods.length > 0 && server.mods.every((m) => set.has(m));
}

/** Launch DayZ and connect to a DayZavr server (loads its mods from !Workshop). */
export async function joinDayzavrServer(server: DayzavrServer) {
  if (!s.profile?.dayzavr_dayz_path) {
    s.setStatus(m.dayzavr_set_path_first(), 'warning');
    return;
  }
  try {
    await invoke('launch_dayzavr_server', {
      host: server.host,
      gamePort: server.game_port,
      password: null,
      mods: server.mods,
    });
    s.setStatus(m.dayzavr_launching({ name: server.name }), 'info');
  } catch (e) {
    s.setStatus(m.dayzavr_launch_failed({ error: String(e) }), 'error');
  }
}

/** Cancel an in-flight DayZavr mod install. */
export async function cancelDayzavrInstall() {
  try {
    await invoke('cancel_dayzavr_install');
  } catch {
    /* non-fatal */
  }
  if (s.dayzavrInstall) s.dayzavrInstall = { ...s.dayzavrInstall, active: false };
}

/** Dismiss the install progress panel. */
export function dismissDayzavrInstall() {
  s.dayzavrInstall = null;
}

/** Remove all installed DayZavr mods from the configured DayZ folder (with confirm). */
export function clearDayzavrMods() {
  const dayzPath = s.profile?.dayzavr_dayz_path;
  if (!dayzPath) {
    s.setStatus(m.dayzavr_set_path_first(), 'warning');
    return;
  }
  s.confirmDialog = {
    title: m.dayzavr_clear_title(),
    message: m.dayzavr_clear_confirm(),
    confirmLabel: m.dayzavr_clear_action(),
    confirmVariant: 'error',
    onConfirm: async () => {
      try {
        const removed = await invoke<string[]>('clear_dayzavr_mods', { dayzPath });
        s.dayzavrInstalledMods = [];
        s.setStatus(m.dayzavr_clear_done({ count: removed.length }), 'success');
      } catch (e) {
        s.setStatus(m.dayzavr_clear_failed({ error: String(e) }), 'error');
      }
    },
  };
}

/** Persist the DayZ install path used for mod installation. */
export async function setDayzavrDayzPath(path: string | null) {
  await saveDayzavrSettings(s.profile?.dayzavr_enabled ?? true, path);
  loadDayzavrInstalledMods();
}

/** Auto-detect the DayZ install via Steam and persist it. Returns the path or null.
 * When `force` is false, does nothing if a path is already configured. */
export async function autodetectDayzPath(force = false): Promise<string | null> {
  if (!force && s.profile?.dayzavr_dayz_path) return s.profile.dayzavr_dayz_path;
  try {
    const path = await invoke<string | null>('detect_dayz_path');
    if (path) await setDayzavrDayzPath(path);
    return path;
  } catch {
    return null;
  }
}

File: dayz-community-hub-ui/src-tauri/src/commands/dayzavr.rs
use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;

use dayz_community_hub_core::api::{Endpoint, Server};
use dayz_community_hub_core::dayzavr::{self, DayzavrServer};
use dayz_community_hub_core::system;
use librqbit::{
    AddTorrent, AddTorrentOptions, AddTorrentResponse, PeerConnectionOptions, Session,
    SessionOptions,
};
use serde::Serialize;
use tauri::ipc::Channel;
use tauri::{AppHandle, State};
use tokio::sync::Mutex;

use crate::helpers::spawn_launch;
use crate::state::{SharedState, insecure_client};

/// Active-install bookkeeping. `gen` is bumped on every new install and on
/// cancel, so a previous install's polling loop can detect it's been superseded
/// and exit cleanly (fixes "can't restart after stopping").
#[derive(Default)]
pub(crate) struct DayzavrInstallInner {
    generation: u64,
    session: Option<Arc<Session>>,
}

pub(crate) struct DayzavrInstallState(pub Mutex<DayzavrInstallInner>);

impl DayzavrInstallState {
    pub(crate) fn new() -> Self {
        Self(Mutex::new(DayzavrInstallInner::default()))
    }
}

/// Progress streamed to the UI during a mod install.
#[derive(Clone, Default, Serialize)]
pub(crate) struct DayzavrInstallProgress {
    pub status: String,
    pub downloaded_bytes: u64,
    pub total_bytes: u64,
    pub uploaded_bytes: u64,
    /// Download speed in MiB/s.
    pub download_mbps: f64,
    /// Upload speed in MiB/s.
    pub upload_mbps: f64,
    /// Connected (live) peers actively transferring.
    pub peers_live: u32,
    /// Peers currently being connected to.
    pub peers_connecting: u32,
    /// Total peers discovered from trackers (sources).
    pub peers_seen: u32,
    /// Human-readable ETA, when known.
    pub eta: Option<String>,
    pub done: bool,
    pub error: Option<String>,
}

fn send_status(ch: &Channel<DayzavrInstallProgress>, status: &str, done: bool, err: Option<String>) {
    let _ = ch.send(DayzavrInstallProgress {
        status: status.to_string(),
        done,
        error: err,
        ..Default::default()
    });
}

/// Download/update only the mods a DayZavr server requires into
/// `<dayz_path>/!Workshop` via the mods-only torrent (selective download).
///
/// Only `@Mod/...` files are fetched (`only_files_regex`); the game-bundle
/// torrent is never touched. Intended for users who own DayZ on Steam.
#[tauri::command]
pub(crate) async fn install_dayzavr_mods(
    mods: Vec<String>,
    dayz_path: String,
    on_progress: Channel<DayzavrInstallProgress>,
    install: State<'_, DayzavrInstallState>,
) -> Result<(), String> {
    let dayz = PathBuf::from(&dayz_path);
    if !dayz.is_dir() {
        return Err(format!("DayZ folder not found: {dayz_path}"));
    }

    // Claim this install: bump the generation and drop any previous session so an
    // older polling loop exits. Our `my_gen` lets us detect if we get superseded.
    let my_gen = {
        let mut g = install.0.lock().await;
        g.generation += 1;
        g.session = None;
        g.generation
    };

    // Always fetch the live manifest and only download mods that are missing or
    // out of date — already-complete mods are skipped entirely. (Partial mods
    // resume: librqbit verifies existing pieces, fetches only what's missing.)
    send_status(&on_progress, "Checking installed mods…", false, None);
    let manifest = dayzavr::fetch_manifest(insecure_client())
        .await
        .map_err(|e| e.to_string())?;
    let to_download = dayzavr::mods_needing_download(&dayz, &manifest, &mods);
    if to_download.is_empty() {
        send_status(&on_progress, "Already up to date", true, None);
        let mut g = install.0.lock().await;
        if g.generation == my_gen {
            g.session = None;
        }
        return Ok(());
    }
    let regex =
        dayzavr::mods_only_regex(&to_download).ok_or_else(|| "No mods to install".to_string())?;

    send_status(&on_progress, "Fetching torrent…", false, None);
    let bytes = insecure_client()
        .get(format!("{}/DayZavr.torrent", dayzavr::MODS_BASE))
        .send()
        .await
        .map_err(|e| e.to_string())?
        .bytes()
        .await
        .map_err(|e| e.to_string())?;

    send_status(&on_progress, "Connecting to swarm…", false, None);
    // Maximise peer discovery + throughput for the 129 GB public swarm:
    // - DHT ON (extra peer source) but persistence OFF (the on-disk DHT state
    //   init is what failed before, not DHT itself).
    // - UPnP ON so peers can also connect inbound (non-fatal if unsupported).
    // - No session persistence (one-shot install).
    // - Short connect timeout so the many dead/unreachable peers in the tracker
    //   set (1000+ "seen", few "live") free their slot fast and we cycle to the
    //   real seeders quickly.
    let session = Session::new_with_opts(
        dayz.clone(),
        SessionOptions {
            disable_dht: false,
            disable_dht_persistence: true,
            persistence: None,
            enable_upnp_port_forwarding: true,
            peer_opts: Some(PeerConnectionOptions {
                connect_timeout: Some(Duration::from_secs(4)),
                read_write_timeout: Some(Duration::from_secs(10)),
                keep_alive_interval: None,
            }),
            ..Default::default()
        },
    )
    .await
    .map_err(|e| e.to_string())?;

    // Store our session only if we're still the current install.
    {
        let mut g = install.0.lock().await;
        if g.generation != my_gen {
            return Ok(()); // superseded before we even started
        }
        g.session = Some(session.clone());
    }

    // `output_folder` and `sub_folder` are mutually exclusive. The session's
    // default output folder is `<dayz>`, so `sub_folder = "!Workshop"` lands the
    // mods at `<dayz>/!Workshop/@Mod/...` (DayZ's mod folder).
    let resp = session
        .add_torrent(
            AddTorrent::from_bytes(bytes.to_vec()),
            Some(AddTorrentOptions {
                only_files_regex: Some(regex),
                sub_folder: Some("!Workshop".to_string()),
                overwrite: true,
                ..Default::default()
            }),
        )
        .await
        .map_err(|e| e.to_string())?;

    let handle = match resp {
        AddTorrentResponse::Added(_, h) => h,
        _ => {
            let mut g = install.0.lock().await;
            if g.generation == my_gen {
                g.session = None;
            }
            return Err("Unexpected torrent response".into());
        }
    };

    loop {
        // Superseded by a new install or cancelled?
        if install.0.lock().await.generation != my_gen {
            send_status(&on_progress, "Cancelled", true, None);
            return Ok(());
        }

        let st = handle.stats();
        let err = st.error.clone();
        let finished = st.finished || err.is_some();
        let live = st.live.as_ref();
        let _ = on_progress.send(DayzavrInstallProgress {
            status: format!("{:?}", st.state),
            downloaded_bytes: st.progress_bytes,
            total_bytes: st.total_bytes,
            uploaded_bytes: st.uploaded_bytes,
            download_mbps: live.map(|l| l.download_speed.mbps).unwrap_or(0.0),
            upload_mbps: live.map(|l| l.upload_speed.mbps).unwrap_or(0.0),
            peers_live: live.map(|l| l.snapshot.peer_stats.live as u32).unwrap_or(0),
            peers_connecting: live
                .map(|l| l.snapshot.peer_stats.connecting as u32)
                .unwrap_or(0),
            peers_seen: live.map(|l| l.snapshot.peer_stats.seen as u32).unwrap_or(0),
            eta: live.and_then(|l| l.time_remaining.as_ref().map(|t| t.to_string())),
            done: finished,
            error: err,
        });

        if finished {
            break;
        }
        tokio::time::sleep(Duration::from_millis(700)).await;
    }

    {
        let mut g = install.0.lock().await;
        if g.generation == my_gen {
            g.session = None;
        }
    }
    Ok(())
}

/// Cancel an in-flight DayZavr mod install (bumps the generation so the polling
/// loop exits and drops the session, stopping the torrent).
#[tauri::command]
pub(crate) async fn cancel_dayzavr_install(
    install: State<'_, DayzavrInstallState>,
) -> Result<(), String> {
    let mut g = install.0.lock().await;
    g.generation += 1;
    g.session = None;
    Ok(())
}

/// Remove all DayZavr-installed mods (folders re-signed with the DayZavr key)
/// from `<dayz_path>/!Workshop`. Returns the removed mod names.
#[tauri::command]
pub(crate) async fn clear_dayzavr_mods(dayz_path: String) -> Result<Vec<String>, String> {
    let dayz = PathBuf::from(&dayz_path);
    dayzavr::clear_installed_mods(&dayz).map_err(|e| e.to_string())
}

/// Names of DayZavr mod folders that are fully installed (complete per the live
/// manifest) under `<dayz_path>/!Workshop`. The UI uses this to only show "Play"
/// when every mod a server requires is present.
#[tauri::command]
pub(crate) async fn dayzavr_installed_mods(dayz_path: String) -> Result<Vec<String>, String> {
    let dayz = PathBuf::from(&dayz_path);
    if !dayz.is_dir() {
        return Ok(vec![]);
    }
    let manifest = dayzavr::fetch_manifest(insecure_client())
        .await
        .map_err(|e| e.to_string())?;
    Ok(dayzavr::installed_complete_mods(&dayz, &manifest))
}

/// Fetch the public DayZavr community server list.
#[tauri::command]
pub(crate) async fn fetch_dayzavr_servers() -> Result<Vec<DayzavrServer>, String> {
    dayzavr::fetch_servers(insecure_client())
        .await
        .map_err(|e| e.to_string())
}

/// Launch DayZ and connect to a DayZavr server, loading the server's mods from
/// `!Workshop` (where `install_dayzavr_mods` placed them). Reuses the normal
/// Steam launch path; the DayZavr mod set is passed as a `-mod=` extra arg.
#[tauri::command]
pub(crate) async fn launch_dayzavr_server(
    host: String,
    game_port: u16,
    password: Option<String>,
    mods: Vec<String>,
    app: AppHandle,
    state: State<'_, SharedState>,
) -> Result<(), String> {
    // DayZ's `-connect` wants an IP; resolve the regional hostname.
    let ip = tokio::net::lookup_host((host.as_str(), game_port))
        .await
        .map_err(|e| format!("Could not resolve {host}: {e}"))?
        .find(std::net::SocketAddr::is_ipv4)
        .map(|a| a.ip().to_string())
        .ok_or_else(|| format!("No address found for {host}"))?;

    let (ctl, dayz_path) = {
        let st = state.read().await;
        (
            st.ctl.clone_for_launch(),
            st.ctl.profile().dayzavr_dayz_path.clone(),
        )
    };

    // Symlink the mods from `!Workshop` into the DayZ root and launch with plain
    // `-mod=@ModName` — exactly how the normal (working) launch loads mods. This
    // avoids the `!Workshop\@ModName` path, which DayZ under Proton often fails to
    // resolve, leaving the client unmodded and BattlEye-kicked.
    let extra: Vec<String> = match (&dayz_path, mods.is_empty()) {
        (_, true) => vec![],
        (Some(p), false) => {
            let available = dayzavr::link_mods_to_root(std::path::Path::new(p), &mods);
            if available.is_empty() {
                return Err("No installed mods found to launch with".into());
            }
            vec![format!("-mod={}", available.join(";"))]
        }
        (None, false) => return Err("DayZ path not set".into()),
    };

    let server = Server {
        endpoint: Endpoint {
            ip,
            port: game_port as i64,
        },
        name: host.clone(),
        game_port: game_port as i64,
        mods: vec![],
        ..Default::default()
    };

    spawn_launch(app, ctl, server, password, extra);
    Ok(())
}

/// Try to auto-detect the DayZ install directory via Steam (libraries, Flatpak,
/// common paths). Returns the path string, or `None` if not found.
#[tauri::command]
pub(crate) async fn detect_dayz_path(state: State<'_, SharedState>) -> Result<Option<String>, String> {
    let hint = {
        let s = state.read().await;
        s.ctl.profile().steam_root.clone()
    };
    Ok(system::detect_dayz_path(hint.as_deref()).map(|p| p.to_string_lossy().into_owned()))
}
